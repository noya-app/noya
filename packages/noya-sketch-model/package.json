{
  "name": "noya-sketch-model",
  "version": "1.0.0",
  "main": "src/index.ts",
  "license": "UNLICENSED"
}
