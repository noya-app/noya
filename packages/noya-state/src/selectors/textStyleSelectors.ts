import Sketch from '@sketch-hq/sketch-file-format-ts';
import { SketchModel } from 'noya-sketch-model';
import { Layers } from 'noya-state';
import { unique } from 'noya-utils';
import { SimpleTextDecoration } from '../primitives';
import { ApplicationState } from '../reducers/applicationReducer';
import { toTextSpans } from './attributedStringSelectors';

export function getTextDecoration(
  encodedAttributes: Pick<
    Sketch.TextStyle['encodedAttributes'],
    'underlineStyle' | 'strikethroughStyle'
  >,
): SimpleTextDecoration {
  return encodedAttributes?.underlineStyle
    ? ('underline' as const)
    : encodedAttributes?.strikethroughStyle
    ? ('strikethrough' as const)
    : ('none' as const);
}

export function getTextStyleAttributes(layer: Sketch.Text) {
  const encodedAttributes = layer.style?.textStyle?.encodedAttributes;
  const paragraphStyle = encodedAttributes?.paragraphStyle;

  const spans = toTextSpans(layer.attributedString);

  const fontNames = [
    ...(encodedAttributes
      ? [encodedAttributes.MSAttributedStringFontAttribute.attributes.name]
      : []),
    ...spans.map(
      (span) => span.attributes.MSAttributedStringFontAttribute.attributes.name,
    ),
  ];

  return {
    fontNames: unique(fontNames),
    fontSize:
      encodedAttributes?.MSAttributedStringFontAttribute.attributes.size ?? 12,
    lineHeight: paragraphStyle?.maximumLineHeight,
    textHorizontalAlignment:
      paragraphStyle?.alignment ?? Sketch.TextHorizontalAlignment.Left,
    textTransform:
      encodedAttributes?.MSAttributedStringTextTransformAttribute ??
      Sketch.TextTransform.None,
    textDecoration: encodedAttributes
      ? getTextDecoration(encodedAttributes)
      : 'none',
  };
}

export function getEncodedStringAttributes(
  style: Sketch.Style | undefined,
): Sketch.StringAttribute['attributes'] {
  const encodedAttributes = style?.textStyle?.encodedAttributes;

  return {
    kerning: encodedAttributes?.kerning,
    textStyleVerticalAlignmentKey:
      encodedAttributes?.textStyleVerticalAlignmentKey,
    MSAttributedStringFontAttribute:
      encodedAttributes?.MSAttributedStringFontAttribute ??
      SketchModel.fontDescriptor(),
    MSAttributedStringColorAttribute:
      encodedAttributes?.MSAttributedStringColorAttribute,
    paragraphStyle: encodedAttributes?.paragraphStyle,
  };
}

export function getAllFontNames(state: ApplicationState) {
  const fontNames = state.sketch.pages.flatMap((page) => {
    const textLayers = Layers.findAll(
      page,
      Layers.isTextLayer,
    ) as Sketch.Text[];

    return textLayers.flatMap((layer) =>
      layer.style?.textStyle
        ? [
            layer.style.textStyle.encodedAttributes
              .MSAttributedStringFontAttribute.attributes.name,
            ...toTextSpans(layer.attributedString).map(
              (span) =>
                span.attributes.MSAttributedStringFontAttribute.attributes.name,
            ),
          ]
        : [],
    );
  });

  return unique(fontNames);
}
