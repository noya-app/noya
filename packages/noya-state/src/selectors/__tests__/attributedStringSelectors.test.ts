import { debugDescription, SketchModel } from 'noya-sketch-model';
import {
  fromTextSpans,
  toTextSpans,
  replaceTextInRange,
  setAttributesInRange,
} from '../attributedStringSelectors';
import { getEncodedStringAttributes } from '../textStyleSelectors';

const empty = SketchModel.attributedString({
  string: '',
  attributes: [],
});

const hello = SketchModel.attributedString({
  string: 'hello',
  attributes: [
    SketchModel.stringAttribute({
      location: 0,
      length: 5,
    }),
  ],
});

const helloWorld = SketchModel.attributedString({
  string: 'helloworld',
  attributes: [
    SketchModel.stringAttribute({
      location: 0,
      length: 5,
    }),
    SketchModel.stringAttribute({
      location: 5,
      length: 5,
    }),
  ],
});

const helloWorld12345 = SketchModel.attributedString({
  string: 'helloworld12345',
  attributes: [
    SketchModel.stringAttribute({
      location: 0,
      length: 5,
    }),
    SketchModel.stringAttribute({
      location: 5,
      length: 5,
    }),
    SketchModel.stringAttribute({
      location: 10,
      length: 5,
    }),
  ],
});

test('convert to and from text spans', () => {
  const spans = toTextSpans(helloWorld);
  const attributedString = fromTextSpans(spans);

  expect(helloWorld).toEqual(attributedString);
});

// Abbreviated to keep `replaceTextInRange` calls on one line
const attrs = getEncodedStringAttributes(undefined);

describe('replace text', () => {
  describe('remove', () => {
    test('one attribute at start', () => {
      const updated = replaceTextInRange(hello, [0, 2], '', attrs);

      expect(debugDescription([hello, updated])).toMatchSnapshot();
    });

    test('one attribute in middle', () => {
      const updated = replaceTextInRange(hello, [2, 3], '', attrs);

      expect(debugDescription([hello, updated])).toMatchSnapshot();
    });

    test('one attribute at end', () => {
      const updated = replaceTextInRange(hello, [3, 5], '', attrs);

      expect(debugDescription([hello, updated])).toMatchSnapshot();
    });

    test('multiple attributes at start', () => {
      const updated = replaceTextInRange(helloWorld, [0, 2], '', attrs);

      expect(debugDescription([helloWorld, updated])).toMatchSnapshot();
    });

    test('multiple attributes in middle', () => {
      const updated = replaceTextInRange(helloWorld, [3, 7], '', attrs);

      expect(debugDescription([helloWorld, updated])).toMatchSnapshot();
    });

    test('multiple attributes at end', () => {
      const updated = replaceTextInRange(helloWorld, [7, 10], '', attrs);

      expect(debugDescription([helloWorld, updated])).toMatchSnapshot();
    });

    test('spanning entire attributes', () => {
      const updated = replaceTextInRange(helloWorld12345, [3, 12], '', attrs);

      expect(debugDescription([helloWorld12345, updated])).toMatchSnapshot();
    });

    test('reversed range', () => {
      const updated = replaceTextInRange(hello, [2, 0], '', attrs);

      expect(debugDescription([hello, updated])).toMatchSnapshot();
    });
  });

  describe('insert', () => {
    test('into empty string', () => {
      const updated = replaceTextInRange(empty, [0, 0], '++', attrs);

      expect(debugDescription([empty, updated])).toMatchSnapshot();
    });

    test('one attribute at start', () => {
      const updated = replaceTextInRange(hello, [0, 2], '++', attrs);

      expect(debugDescription([hello, updated])).toMatchSnapshot();
    });

    test('one attribute in middle', () => {
      const updated = replaceTextInRange(hello, [2, 3], '++', attrs);

      expect(debugDescription([hello, updated])).toMatchSnapshot();
    });

    test('one attribute at end', () => {
      const updated = replaceTextInRange(hello, [3, 5], '++', attrs);

      expect(debugDescription([hello, updated])).toMatchSnapshot();
    });

    test('multiple attributes at start', () => {
      const updated = replaceTextInRange(helloWorld, [0, 2], '++', attrs);

      expect(debugDescription([helloWorld, updated])).toMatchSnapshot();
    });

    test('multiple attributes in middle', () => {
      const updated = replaceTextInRange(helloWorld, [3, 7], '++', attrs);

      expect(debugDescription([helloWorld, updated])).toMatchSnapshot();
    });

    test('multiple attributes at end', () => {
      const updated = replaceTextInRange(helloWorld, [7, 10], '++', attrs);

      expect(debugDescription([helloWorld, updated])).toMatchSnapshot();
    });

    test('spanning entire attributes', () => {
      const updated = replaceTextInRange(helloWorld12345, [3, 12], '++', attrs);

      expect(debugDescription([helloWorld12345, updated])).toMatchSnapshot();
    });

    test('beyond current string', () => {
      const updated = replaceTextInRange(hello, [10, 10], '++', attrs);

      expect(debugDescription([hello, updated])).toMatchSnapshot();
    });
  });
});

describe('set attributes', () => {
  const attrs = {
    MSAttributedStringColorAttribute: SketchModel.WHITE,
  };
  const options = {
    attributedStringColors: true,
  };

  test('into empty string', () => {
    const updated = setAttributesInRange(empty, [0, 0], attrs);

    expect(debugDescription([empty, updated], options)).toMatchSnapshot();
  });

  test('one attribute at start', () => {
    const updated = setAttributesInRange(hello, [0, 2], attrs);

    expect(debugDescription([hello, updated], options)).toMatchSnapshot();
  });

  test('one attribute in middle', () => {
    const updated = setAttributesInRange(hello, [2, 3], attrs);

    expect(debugDescription([hello, updated], options)).toMatchSnapshot();
  });

  test('one attribute at end', () => {
    const updated = setAttributesInRange(hello, [3, 5], attrs);

    expect(debugDescription([hello, updated], options)).toMatchSnapshot();
  });

  test('multiple attributes at start', () => {
    const updated = setAttributesInRange(helloWorld, [0, 2], attrs);

    expect(debugDescription([helloWorld, updated], options)).toMatchSnapshot();
  });

  test('multiple attributes in middle', () => {
    const updated = setAttributesInRange(helloWorld, [3, 7], attrs);

    expect(debugDescription([helloWorld, updated], options)).toMatchSnapshot();
  });

  test('multiple attributes at end', () => {
    const updated = setAttributesInRange(helloWorld, [7, 10], attrs);

    expect(debugDescription([helloWorld, updated], options)).toMatchSnapshot();
  });

  test('spanning entire attributes', () => {
    const updated = setAttributesInRange(helloWorld12345, [3, 12], attrs);

    expect(
      debugDescription([helloWorld12345, updated], options),
    ).toMatchSnapshot();
  });
});
