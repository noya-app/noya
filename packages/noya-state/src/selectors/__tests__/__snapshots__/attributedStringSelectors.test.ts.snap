// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replace text insert beyond current string 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5 }

<AttributedString> \\"hello++\\"
└── <StringAttribute> { location: 0, length: 7 }"
`;

exports[`replace text insert into empty string 1`] = `
"<AttributedString> \\"\\"

<AttributedString> \\"++\\"
└── <StringAttribute> { location: 0, length: 2 }"
`;

exports[`replace text insert multiple attributes at end 1`] = `
"<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 5 }

<AttributedString> \\"hellowo++\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 4 }"
`;

exports[`replace text insert multiple attributes at start 1`] = `
"<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 5 }

<AttributedString> \\"++lloworld\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 5 }"
`;

exports[`replace text insert multiple attributes in middle 1`] = `
"<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 5 }

<AttributedString> \\"hel++rld\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 3 }"
`;

exports[`replace text insert one attribute at end 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5 }

<AttributedString> \\"hel++\\"
└── <StringAttribute> { location: 0, length: 5 }"
`;

exports[`replace text insert one attribute at start 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5 }

<AttributedString> \\"++llo\\"
└── <StringAttribute> { location: 0, length: 5 }"
`;

exports[`replace text insert one attribute in middle 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5 }

<AttributedString> \\"he++lo\\"
└── <StringAttribute> { location: 0, length: 6 }"
`;

exports[`replace text insert spanning entire attributes 1`] = `
"<AttributedString> \\"helloworld12345\\"
├── <StringAttribute> { location: 0, length: 5 }
├── <StringAttribute> { location: 5, length: 5 }
└── <StringAttribute> { location: 10, length: 5 }

<AttributedString> \\"hel++345\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 3 }"
`;

exports[`replace text remove multiple attributes at end 1`] = `
"<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 5 }

<AttributedString> \\"hellowo\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 2 }"
`;

exports[`replace text remove multiple attributes at start 1`] = `
"<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 5 }

<AttributedString> \\"lloworld\\"
├── <StringAttribute> { location: 0, length: 3 }
└── <StringAttribute> { location: 3, length: 5 }"
`;

exports[`replace text remove multiple attributes in middle 1`] = `
"<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5 }
└── <StringAttribute> { location: 5, length: 5 }

<AttributedString> \\"helrld\\"
├── <StringAttribute> { location: 0, length: 3 }
└── <StringAttribute> { location: 3, length: 3 }"
`;

exports[`replace text remove one attribute at end 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5 }

<AttributedString> \\"hel\\"
└── <StringAttribute> { location: 0, length: 3 }"
`;

exports[`replace text remove one attribute at start 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5 }

<AttributedString> \\"llo\\"
└── <StringAttribute> { location: 0, length: 3 }"
`;

exports[`replace text remove one attribute in middle 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5 }

<AttributedString> \\"helo\\"
└── <StringAttribute> { location: 0, length: 4 }"
`;

exports[`replace text remove reversed range 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5 }

<AttributedString> \\"llo\\"
└── <StringAttribute> { location: 0, length: 3 }"
`;

exports[`replace text remove spanning entire attributes 1`] = `
"<AttributedString> \\"helloworld12345\\"
├── <StringAttribute> { location: 0, length: 5 }
├── <StringAttribute> { location: 5, length: 5 }
└── <StringAttribute> { location: 10, length: 5 }

<AttributedString> \\"hel345\\"
├── <StringAttribute> { location: 0, length: 3 }
└── <StringAttribute> { location: 3, length: 3 }"
`;

exports[`set attributes into empty string 1`] = `
"<AttributedString> \\"\\"

<AttributedString> \\"\\""
`;

exports[`set attributes multiple attributes at end 1`] = `
"<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 5, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }

<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
├── <StringAttribute> { location: 5, length: 2, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 7, length: 3, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }"
`;

exports[`set attributes multiple attributes at start 1`] = `
"<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 5, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }

<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 2, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }
├── <StringAttribute> { location: 2, length: 3, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 5, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }"
`;

exports[`set attributes multiple attributes in middle 1`] = `
"<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 5, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }

<AttributedString> \\"helloworld\\"
├── <StringAttribute> { location: 0, length: 3, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
├── <StringAttribute> { location: 3, length: 2, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }
├── <StringAttribute> { location: 5, length: 2, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 7, length: 3, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }"
`;

exports[`set attributes one attribute at end 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }

<AttributedString> \\"hello\\"
├── <StringAttribute> { location: 0, length: 3, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 3, length: 2, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }"
`;

exports[`set attributes one attribute at start 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }

<AttributedString> \\"hello\\"
├── <StringAttribute> { location: 0, length: 2, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 2, length: 3, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }"
`;

exports[`set attributes one attribute in middle 1`] = `
"<AttributedString> \\"hello\\"
└── <StringAttribute> { location: 0, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }

<AttributedString> \\"hello\\"
├── <StringAttribute> { location: 0, length: 2, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
├── <StringAttribute> { location: 2, length: 1, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 3, length: 2, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }"
`;

exports[`set attributes spanning entire attributes 1`] = `
"<AttributedString> \\"helloworld12345\\"
├── <StringAttribute> { location: 0, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
├── <StringAttribute> { location: 5, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 10, length: 5, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }

<AttributedString> \\"helloworld12345\\"
├── <StringAttribute> { location: 0, length: 3, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }
├── <StringAttribute> { location: 3, length: 2, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }
├── <StringAttribute> { location: 5, length: 5, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }
├── <StringAttribute> { location: 10, length: 2, color: { red: 1, green: 1, blue: 1, alpha: 1, _class: \\"color\\" } }
└── <StringAttribute> { location: 12, length: 3, color: { red: 0, green: 0, blue: 0, alpha: 1, _class: \\"color\\" } }"
`;
